File: Ant/chapter2/python/counting_lattice_points_on_a_line_segment.py
def ns(f):
    return next(f).strip()


with open("../testset/counting_lattice_points_on_a_line_segment/test1.txt", 'r') as f:
    x1, y1 = map(int, ns(f).split())
    x2, y2 = map(int, ns(f).split())


def gcd(a, b):
    if b == 0:
        return a
    return gcd(b, a % b)


a = abs(x1 - x2)
b = abs(y1 - y2)
ans = gcd(a, b) - 1
print(ans)

File: Ant/chapter2/cpp/counting_lattice_points_on_a_line_segment.cpp
#include <fstream>
#include <math.h>
#include <stdio.h>

using namespace std;

int X1, X2, Y1, Y2, ans;

int gcd(int a, int b) {
  if (b == 0) {
    return a;
  }
  return gcd(b, a % b);
}

int main() {
  ifstream ifs("../testset/counting_lattice_points_on_a_line_segment/test1.txt");
  ifs >> X1 >> Y1 >> X2 >> Y2;
  int a = abs(X1 - X2);
  int b = abs(Y1 - Y2);
  ans = gcd(a, b) - 1;
  printf("%d\n", ans);
}